presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java

```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
                 StreamDescriptor streamDescriptor = streamDescriptors.get(columnId);
+
+                Type type = includedColumns.get(columnId);
+                if (!validateOrcType(streamDescriptor, type.getTypeSignature())) {
+                    throw new OrcCorruptionException("ORC type from file footer (%s) does not match expected type of table (%s)", types, includedColumns);
+                }
+
                 streamReaders[columnId] = StreamReaders.createStreamReader(streamDescriptor, hiveStorageTimeZone);
             }
         }
         return streamReaders;
     }
 
+    private static boolean validateOrcType(StreamDescriptor streamDescriptor, TypeSignature type)
+    {
+        OrcTypeKind orcTypeKind = streamDescriptor.getStreamType();
+        List<StreamDescriptor> streamDescriptorList = streamDescriptor.getNestedStreams();
+        List<TypeSignatureParameter> typeSignatureParameterList = type.getParameters();
+        switch (type.getBase()) {
+            case StandardTypes.VARCHAR:
+                return (orcTypeKind == OrcTypeKind.STRING || orcTypeKind == OrcTypeKind.VARCHAR || orcTypeKind == OrcTypeKind.CHAR) &&
+                        streamDescriptorList.size() == 0 &&
+                        typeSignatureParameterList.size() == 1;
+            case StandardTypes.ARRAY:
+                return orcTypeKind == OrcTypeKind.LIST &&
+                        streamDescriptorList.size() == 1 &&
+                        typeSignatureParameterList.size() == 1 &&
+                        validateOrcType(streamDescriptorList.get(0), typeSignatureParameterList.get(0).getTypeSignature());
+            case StandardTypes.MAP:
+                return orcTypeKind == OrcTypeKind.MAP &&
+                        streamDescriptorList.size() == 2 &&
+                        typeSignatureParameterList.size() == 2 &&
+                        validateOrcType(streamDescriptorList.get(0), typeSignatureParameterList.get(0).getTypeSignature()) &&
+                        validateOrcType(streamDescriptorList.get(1), typeSignatureParameterList.get(1).getTypeSignature());
+            case StandardTypes.ROW:
+                if (orcTypeKind != OrcTypeKind.STRUCT || streamDescriptorList.size() != typeSignatureParameterList.size()) {
+                    return false;
+                }
+                for (int i = 0; i < streamDescriptorList.size(); ++i) {
+                    if (!validateOrcType(streamDescriptorList.get(i), typeSignatureParameterList.get(i).getNamedTypeSignature().getTypeSignature())) {
+                        return false;
+                    }
+                }
+                return true;
+            case StandardTypes.BOOLEAN:
+                return orcTypeKind == OrcTypeKind.BOOLEAN;
+            case StandardTypes.DOUBLE:
+                return orcTypeKind == OrcTypeKind.FLOAT ||
+                        orcTypeKind == OrcTypeKind.DOUBLE;
+            case StandardTypes.BIGINT:
```

Comment:
This might need to be updated for the newly added integer types. cc @raghavsethi